### DISCUSSION TITLE
Audiokit , how to make sd player go a2dp source ??

### DISCUSSION THREAD

**Discussion by TangerineDreamer** [pschatzmann/arduino-audio-tools#364] [Q&A] [answered]
Hi, i'm sruggling with it today,there is no examples with SD player and A2P source,
only with A2P SINK,here is my sketch try:

/--------------------------------------------------------

#define USE_SDFAT
#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioA2DP.h"

const char *startFilePath="/";
const char* ext="mp3";
BluetoothA2DPSource a2dp_source;

SdSpiConfig sdcfg(PIN_AUDIO_KIT_SD_CARD_CS, DEDICATED_SPI, SD_SCK_MHZ(10) , &AUDIOKIT_SD_SPI);
AudioSourceSDFAT source(startFilePath, ext, sdcfg);
MP3DecoderHelix decoder;  

AudioKitStream i2s;
AudioPlayer player(source, i2s, decoder);

const int16_t BYTES_PER_FRAME = 4;

// callback used by A2DP to provide the sound data - usually len is 128 2 channel int16 frames
int32_t get_sound_data(Frame* data, int32_t frameCount) {
  return i2s.readBytes((uint8_t*)data, frameCount*BYTES_PER_FRAME)/BYTES_PER_FRAME;
}

void next(bool, int, void*) {
   player.next();
}

void previous(bool, int, void*) {
   player.previous();
}

  void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  // start i2s input with default configuration
  Serial.println("starting I2S...");
  auto cfg = i2s.defaultConfig(RXTX_MODE);
  cfg.sd_active = true;

  

  
  cfg.bits_per_sample = 16;
  cfg.channels = 2;
  cfg.sample_rate = 44100;
  
  i2s.begin(cfg);

  // setup additional buttons 
  i2s.addAction(PIN_KEY4, next);
  i2s.addAction(PIN_KEY3, previous);

   // setup player
  player.setVolume(0.7);
  player.begin();

   // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.start("ESPAT Musik", get_sound_data);  
}

void loop() {
  player.copy();
  i2s.processActions();
}

//--------------------------------------------------------------

I guess that " player.copy();" does not make the deal
I was more easy to make line input go to a2dp source.
Mr shatzmann,how do you do ?......(the sketch of course) :)

**Comment by pschatzmann** [comment, marked as answer]
What's wrong with this example: https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-player/player-sdfat-a2dp ?

**Comment by TangerineDreamer** [comment]
That's make my audiokit reboot after connecting,here is the log:

[W] AudioA2DP.h : 302 - ==> state: Connecting
[W] AudioA2DP.h : 302 - ==> state: Connected
[E] AudioSourceSDFAT.h : 89 - sd.begin failed
[E] SDDirect.h : 231 - Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

Core  1 register dump:
PC      : 0x40091531  PS      : 0x00060330  A0      : 0x8018d56c  A1      : 0x3ffcafb0  
A2      : 0x00000000  A3      : 0xfffffffc  A4      : 0x000000ff  A5      : 0x0000ff00  
A6      : 0x00ff0000  A7      : 0xff000000  A8      : 0x00000000  A9      : 0x3ffcb280  
A10     : 0x00000003  A11     : 0x00060323  A12     : 0x00060320  A13     : 0x0000ff00  
A14     : 0x00ff0000  A15     : 0xff000000  SAR     : 0x0000001b  EXCCAUSE: 0x0000001c  
EXCVADDR: 0x00000000  LBEG    : 0x40091531  LEND    : 0x40091541  LCOUNT  : 0xffffffff  


Backtrace:0x4009152e:0x3ffcafb00x4018d569:0x3ffcafc0 0x4018a6f5:0x3ffcb2d0 0x400d65c7:0x3ffcb390 0x400d6b96:0x3ffcb3b0 0x400d6f72:0x3ffcb480 0x400d5e08:0x3ffcb4a0 0x400d5f39:0x3ffcb4c0 0x400e0cca:0x3ffcb520 




ELF file SHA256: 0000000000000000

Rebooting...

**Comment by pschatzmann** [comment]
Did you replace the I2SStream with the AudioKitStream and activate the SD correctly ?
https://github.com/pschatzmann/arduino-audio-tools/wiki/The-Audio-Player-Class
https://github.com/pschatzmann/arduino-audiokit/wiki/Problems-with-the-SD-Drive-or-some-conflicting-Buttons

And please learn how to decode the stacktrace: https://github.com/me-no-dev/EspExceptionDecoder

**Comment by TangerineDreamer** [comment]
Now the sd card is ok,i can hear it in the headphones,but not via the bluetooth,only the 2 mics
are passing on the air,but i don't want them,i've no line input while the mics are hearded.
here is my code:

//--------------------------------------------------------------

/**
 * @file player-sdfat-a2dp.ino
 * @brief see https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-player/player-sdfat-a2dp/README.md
 * 
 * @author Phil Schatzmann
 * @copyright GPLv3
 */


#include "AudioTools.h"
#include "AudioLibs/AudioKit.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSDFAT.h" 
#include "AudioCodecs/CodecMP3Helix.h"

BluetoothA2DPSource a2dp_source;

const char *startFilePath="/";
const char* ext="mp3";
SdSpiConfig sdcfg(PIN_AUDIO_KIT_SD_CARD_CS, DEDICATED_SPI, SD_SCK_MHZ(10) , &AUDIOKIT_SD_SPI);
AudioSourceSDFAT source(startFilePath, ext, sdcfg);

AudioKitStream i2s;
MP3DecoderHelix decoder;

const int16_t BYTES_PER_FRAME = 4;

// callback used by A2DP to provide the sound data - usually len is 128 2 channel int16 frames
int32_t get_sound_data(Frame* data, int32_t frameCount) {
  return i2s.readBytes((uint8_t*)data, frameCount*BYTES_PER_FRAME)/BYTES_PER_FRAME;
}


AudioPlayer player(source,i2s, decoder);


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);  
    
  
    
  auto cfg = i2s.defaultConfig(RXTX_MODE);
  cfg.bits_per_sample = 16;
  cfg.channels = 2;
  cfg.sample_rate = 44100;
  
  cfg.sd_active = true;
  i2s.begin(cfg);

   // setup player
  player.setVolume(0.8);
  player.begin();

  // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.start("My Musik", get_sound_data);  

}

void loop() {
  player.copy();
}

//--------------------------------------------------------------------

I have no really errors in the monitor log:

//---------------------------------------------------------
[W] AudioKit.h : 51 - sd_active = true
[W] AudioKit.h : 602 - Mode Button ignored because of conflict: 13 
[W] AudioKit.h : 610 - Headphone detection ignored because of conflict: 5 
starting A2DP...
[W] AudioPlayer.h : 309 - -> timeout - moving to next stream
[W] AudioPlayer.h : 309 - -> timeout - moving to next stream
[W] AudioPlayer.h : 309 - -> timeout - moving to next stream
//---------------------------------------------------------------------------

my switches are 0-1-1-0-0, board-->(6)

**Comment by pschatzmann** [comment]
Forget my comment about replacing the I2SStream with the AudioKitStream. I got confused...
Your sketch above is a big mess and I don't understand what you are doing: In the initial issue you write that you want the Audio Player to play to A2DP but in you sketch you copy the microphone data to A2DP and use the Audio Player at the same time!

Just use the sketch that I proposed above. The only issue with this is that, since you don't need any AudioKitStream, the SPI does not get initialized and you need to do it yourself.

I added a [section to the documentation](https://github.com/pschatzmann/arduino-audiokit/wiki/Problems-with-the-SD-Drive-or-some-conflicting-Buttons) how to do this. Or alternatively use SDMMC...

**Comment by TangerineDreamer** [comment]
But i had SD issues with the sketch you suggested,not with my last one,i only need to synthetize in mind  the use of those functions and
i believe it's good to make it more clear:

// callback used by A2DP to provide the sound data - usually len is 128 2 channel int16 frames
int32_t get_sound_data(Frame* data, int32_t frameCount) {
return i2s.readBytes((uint8_t*)data, frameCount*BYTES_PER_FRAME)/BYTES_PER_FRAME;
}

As i try to understand,A2DP take in the frames ,like a streamcopy by frame ,what going out of the audiokit (i2s),and it's not
by using the copier function in the loop,am i right?


AudioPlayer player(source,i2s, decoder);
Now i think the audio processing take out (the source : SD) ,via the (decoder), to headphones (i2s), what is going in the audiokit,i think the SD player is natively connected to the input of the audiokit,and i can change by setting the INPUT:INPUT_LINE2 (or something like that),
but the AudioPlayer wait for SD data but have nothing,so it switches to the next input stream : the MICS,and that's why i hear them in
the bluetooth speaker am i right?

Then i tried to write : AudioPlayer player(source, A2DP_source, decoder); and it said AudioPlayer class don't allow that.
All i want ,is to connect the SD player of the audiokit  to A2DP,but there is no INPUT_SD define in the settings.
 

I'm sure i'm not alone to be confused with all of that :)

**Comment by pschatzmann** [comment]
int32_t get_sound_data(Frame* data, int32_t frameCount) {
return i2s.readBytes((uint8_t*)data, frameCount*BYTES_PER_FRAME)/BYTES_PER_FRAME;
}
Is reading the data from the Audiokit Microphone using the A2DP Callback directly.
The AudioPlayer is asking for an Audio Sink and you provide I2S (which is the amplifier of the audiokit), but you need to have a Stream that sends the data to A2DP...

All you need to do is take [this example](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-player/player-sdfat-a2dp) and make sure that the SD is working, by initializing SPI or by using SDMMC

As an excercise it might be even helpful if you try create a learning test sketch which opens and reads a file!

**Comment by TangerineDreamer** [comment]
That's not the SD that makes problem,i've incorporate the SD MMC in the sketch and the result is ok.
Now it says : connecting then disconnected all the time without to be connected for sure.
What's the difference between AudioKitHAL.h and AudioKit.h?

**Comment by pschatzmann** [comment]
Maybe https://github.com/pschatzmann/ESP32-A2DP/wiki/Auto-Reconnect helps. 
Connecting might take quite some time. Make sure that no other device is connected: sometimes this happens automatically!

AudioKit.h is the stream API from the Audio Tools which is using AudioKitHAL.h which is part of the AudioKit HAL Project

**Comment by TangerineDreamer** [comment]
a2dp_source.set_auto_reconnect(true) does not work,same for : out.set_auto_reconnect(true)
it's not recognized.
before i tried: cfg2.auto_reconnect = true;,no problem for compiling but the problem is told above.

**Comment by pschatzmann** [comment]
The issue is that is might be set to true: Change it to __false__!
Did you deactivate the bluetooth of all other devices that might connect ?
What device are you trying to connect to ?

**Comment by TangerineDreamer** [comment]
i have a esp32 wrover near at 10 cm from the sender, my pc is windows 7 ,no bluetooth i believe,
it's weird because all my other sketches worked fine with bluetooth.
The example you suggested is with a single esp32 and an external SD player,surely i can do it with
success,but did you try and succeed what i'm doing with the internal audiokit SD player ?

**Comment by pschatzmann** [comment]
The problem is not with your audiokit but with your bluetooth speaker that does not accept any connect request. If it is not working with an audiokit it will not work with an ESP32 as well. 

I always recommend to start simple: so to test A2DP you should have used the https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-stream/streams-generator-a2dp sketch first...

Did you try to restart the Bluetooth Speaker. Maybe it is just hanging...
You can also activate the ESP32 logging in Arduino to see what's going one in detail.

**Comment by TangerineDreamer** [comment]
Oh,it's not a real bluetooth speaker,only a esp32 receiver ,with a basic sketch from your's:

//-------------------------------------------------------

#include "BluetoothA2DPSink.h"

BluetoothA2DPSink a2dp_sink;

void setup() {

//---------PIN CONFIG------------  
    i2s_pin_config_t my_pin_config = {
        .bck_io_num = 26,
        .ws_io_num = 25,
        .data_out_num = 22,
        .data_in_num = I2S_PIN_NO_CHANGE
    };
    a2dp_sink.set_pin_config(my_pin_config);
//-------------------------------------
    
//------------I2S CONFIG---------------
static i2s_config_t i2s_config = {
      .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_TX),
      .sample_rate = 44100, // updated automatically by A2DP
      .bits_per_sample = (i2s_bits_per_sample_t)16,
      .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
      .communication_format = (i2s_comm_format_t) (I2S_COMM_FORMAT_STAND_I2S),
      .intr_alloc_flags = 0, // default interrupt priority
      .dma_buf_count = 8,
      .dma_buf_len = 64,
      .use_apll = true,
      .tx_desc_auto_clear = true // avoiding noise in case of data unavailability
  };
  a2dp_sink.set_i2s_config(i2s_config);
//-----------------------------------------
    
    a2dp_sink.start("My Musik");
}

void loop() {
}

//---------------------------------------------------------------------

**Comment by pschatzmann** [comment]
Hmm, that should work as well: Make sure on that sketch that the auto reconnect is as well disabled!

**Comment by TangerineDreamer** [comment]
Yes,it has never been on the sketch,i show you the sketch of the sender with the sd mmc implant:

//----------------------------------------------------------

#include "AudioKitHAL.h"
#include "SD_MMC.h"
#include "AudioTools.h"
#include "AudioLibs/AudioA2DP.h"
#include "AudioLibs/AudioSourceSDFAT.h"
#include "AudioCodecs/CodecMP3Helix.h"

const char *startFilePath="/";
const char* ext="mp3";
AudioSourceSDFAT source(startFilePath, ext);
A2DPStream out = A2DPStream::instance();  // A2DP input - A2DPStream is a singleton!
MP3DecoderHelix decoder;
AudioPlayer player(source, out, decoder);

AudioKit kit;

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning);

  auto cfg = kit.defaultConfig();
    cfg.sd_active = false;
    kit.begin(cfg);
    if(!SD_MMC.begin("/sdcard", true)){    // 1-bit mode
        Serial.println("Card Mount Failed");
        return;
    } else {
        Serial.println("Card Mount Success");
    }

  // setup output - We send the test signal via A2DP - so we conect to the "LEXON MINO L" Bluetooth Speaker
  auto cfg2 = out.defaultConfig(TX_MODE);
  cfg2.name = "My Musik";
  cfg2.auto_reconnect = false;  // if this is use we just quickly connect to the last device ignoring cfg.name
  
  out.begin(cfg2);

  // setup player
  player.setVolume(0.6);
  player.begin();

}

void loop() {
  player.copy();
}

//-----------------------------------------------------------------

**Reply by pschatzmann** [reply]
This is still messed up: You can't mix AudioSourceSDFAT with SDMMC: You need to be consistent!
See the examples in [AudioSourceSDFAT](https://github.com/pschatzmann/arduino-audio-tools/wiki/The-Audio-Player-Class).

I suggest that you make one step back and try the simple sine generator. If this is working add the file support! Otherwise your issues are all over the place and there is no chance to figure out what's working and what not!

**Reply by TangerineDreamer** [reply]
You're right !!,i just put instead  #include "AudioLibs/AudioSourceSDMMC.h" ,
and  AudioSourceSDMMC source(startFilePath, ext);
now i'm connected,it's a good point,but it say

[W] AudioA2DP.h : 302 - ==> state: Connecting
[W] AudioA2DP.h : 302 - ==> state: Connected
[W] AudioPlayer.h : 141 - -> begin: no data found

i have a bunch of mp3 inside and i always could read them in the audiokit sketches

**Reply by pschatzmann** [reply]
Did you test the SDMMC with output to AudioKitStream ? I guess you still have not everything right.
Are all dip switches in the on position ?
Did you restart your board after changing the dip switches ?

**Comment by TangerineDreamer** [comment]
i have change the switches to All ON and restarted the board wich is the 5) option one ,but it's the
same.
I can tell you that the generator was alright with bluetooth,the mics also,only the sd player does
not give something,but worked in sdfat to feed the headphones

**Reply by pschatzmann** [reply]
You need to make a choice: The statement that SDFAT was working is useless if you want SDMMC to work.
Did you test the SDMMC with output to AudioKitStream: You just need to run this sketch: https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sdmmc-audiokit/player-sdmmc-audiokit.ino

**Comment by TangerineDreamer** [comment]
Seems the same issue : no data found using the mp3 it reconized,all the switches are on, board(5)

[I] AudioTypes.h : 53 - sample_rate: 44100
[I] AudioTypes.h : 54 - channels: 2
[I] AudioTypes.h : 55 - bits_per_sample: 16
[I] I2SConfig.h : 95 - rx/tx mode: TX
[I] I2SConfig.h : 96 - port_no: 0
[I] I2SConfig.h : 97 - is_master: Master
[I] I2SConfig.h : 98 - sample rate: 44100
[I] I2SConfig.h : 99 - bits per sample: 16
[I] I2SConfig.h : 100 - number of channels: 2
[I] I2SConfig.h : 101 - i2s_format: I2S_STD_FORMAT
[I] I2SConfig.h : 103 - auto_clear:0
[I] I2SConfig.h : 110 - buffer_count:10
[I] I2SConfig.h : 111 - buffer_size:512
[I] I2SConfig.h : 115 - pin_bck: 14
[I] I2SConfig.h : 116 - pin_ws: 15
[I] I2SConfig.h : 117 - pin_data: -1515870811
[W] AudioKit.h : 51 - sd_active = false
[I] AudioKit.h : 199 - Channels 2->2
[I] AudioKit.h : 590 - void audio_tools::AudioKitStream::setupActions()
[W] AudioKit.h : 593 - Deactivating SPI because SD is not active
[I] AudioKit.h : 340 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 13 / logic: 0
[I] AudioActions.h : 52 - pin 13 -> INPUT_PULLUP
[I] AudioActions.h : 42 - ActionLogic::add pin: 39 / logic: 2
[I] AudioActions.h : 55 - pin 39 -> INPUT
[I] AudioKit.h : 340 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 18 / logic: 0
[I] AudioActions.h : 52 - pin 18 -> INPUT_PULLUP
[I] AudioKit.h : 340 - void audio_tools::AudioKitStream::addAction(int, void (*)(bool, int, void*), void*)
[I] AudioActions.h : 42 - ActionLogic::add pin: 5 / logic: 0
[I] AudioActions.h : 52 - pin 5 -> INPUT_PULLUP
[I] AudioPlayer.h : 272 - setVolume(0.700000)
[I] AudioStreams.h : 1104 - setVolume: 0.700000
[I] AudioStreams.h : 1104 - setVolume: 0.700000
[I] MetaDataID3.h : 545 - virtual void audio_tools::MetaDataID3::begin()
[I] AudioSourceSDMMC.h : 62 - selectStream: 0
[I] AudioSourceSDMMC.h : 66 - Using file Ain't That A Mother's Luck.mp3
[W] AudioPlayer.h : 141 - -> begin: no data found

**Comment by TangerineDreamer** [comment]
player sd fat audiokit is the only one that works for me.

**Comment by pschatzmann** [comment]
Can it be that you did not use the official SD Formatter ?
If you have any issues with the formatting I suggest to use the [official SD Formatter](https://www.sdcard.org/downloads/formatter/)

**Comment by TangerineDreamer** [comment]
First,i thank you for your support :)
I'm sorry to say that the issues : "no data found" are the same as before,even the sd recorder gives 0k rec file,but the mp3 names are recognized.
Wtch audiokit d you use?
I use the board 5) and the 6),i've got the lyra too,give me exactly the procedure to reproduce what
you succeeded.

**Comment by pschatzmann** [comment]
There should not be any difference: all boards are wired the same!
Please confirm that you have reformatted the SD drive with the official formatter and that this did not resolve your issue. I assume that you did your tests all with the same SD card.

I just did a quick retest with the SDFAT example and it is still working properly for me and I remember reading that peaple that formatted the SD drive on their own PC had the same kind of issues...

**Comment by TangerineDreamer** [comment]
Yes i did reformat with the official formater,
Now i have another direction to look for,if the mp3 names are recognized it's certainly not the
sd fault,it says it could not find the data after it wrote the mp3 name,so let's try to find what data
it's looking for,are there any mp3 limitations concerning the bitrate or thoses limitations are they
in the AudioSourceSdXX.h?
I know they aren't in the helixcodec because the AudioSourceSdfat. let pass the data and i can hear.

**Comment by pschatzmann** [comment]
Strange: You get the message 'no data found' if the AudioPlayer receives nullptr instead of a valid file.
Try to do a file read sketch just with the file API w/o my framework and you should be able to reproduce the issue.
With SDFAT you can also try to reduce the read speed!
Then it is better to log the issue with the corresponding library...
I think the SDFAT library has some test sketches...

**Comment by TangerineDreamer** [comment]
sorry,i'm not a programmer veteran,can you explain more,are you talking about another mp3 player?

**Comment by pschatzmann** [comment]
The SDFAT examples are using [this library](https://github.com/greiman/SdFat)
The parameters are 
SdSpiConfig sdcfg(PIN_AUDIO_KIT_SD_CARD_CS, DEDICATED_SPI, SD_SCK_MHZ(10) , &AUDIOKIT_SD_SPI);

so you can try to decrease the SD_SCK_MHZ(10) to e.g. SD_SCK_MHZ(2)

I am starting to get confused: which examples are not working ?
- sdfat
- sd
- sdmmc

**Comment by TangerineDreamer** [comment]
Ok,i uploaded the new sdfat library,seems i had not the version2,
So to remain all,the sdfat player was ok,the sdplayer had issue,same for sdmmc player

**Comment by TangerineDreamer** [comment]
Have you seen that?
Those hacker box guy also have issues with you library or the audiokit one:

https://www.instructables.com/HackerBox-0079-Audio-DSP/

**Comment by pschatzmann** [comment]
Just tried the SD library with the latest ESP32 version and I can confirm that this is failing...

**Comment by TangerineDreamer** [comment]
Ok,that's maybe a solution,i've the esp32 package 2.04 and it's a bunch of bugs to soluce.
i will downgrade to 2.02,or the version you have,wich one do you have?

**Comment by pschatzmann** [comment]
I suggest that you stick with the SDFAT library until I have a better visibility...

**Comment by TangerineDreamer** [comment]
My mp3's are working nicely and smoothly with the audiokit sdfat player,what example of the sdfat
library do you want me to test,because they are many.

**Comment by TangerineDreamer** [comment]
the esp32 package v.2.05 is out,i'm gonna try it

**Comment by TangerineDreamer** [comment]
they made some change in sd libraries that could interest you,they also repared espnow channel bug.
https://github.com/espressif/arduino-esp32/tree/master/libraries

**Comment by TangerineDreamer** [comment]
bad thing,it can't compile because it makes a very very long text in the verbose monitor and makes the memory full,i'm back to downgrade to 2.01,'cause the 2.02 disapeared

**Comment by pschatzmann** [comment]
I did some tests with the audiokit with sdfat -> a2dp, but there are quite a few issues there as well, that I need to look into..

**Comment by TangerineDreamer** [comment]
I can believe you, i have :  sd.begin failed after connected in "player-sdfat-a2dp"

[W] AudioA2DP.h : 302 - ==> state: Connecting
[W] AudioA2DP.h : 302 - ==> state: Connected
[E] AudioSourceSDFAT.h : 89 - sd.begin failed
[E] SDDirect.h : 231 - Guru Meditation Error: Core  1 panic'ed (LoadProhibited). Exception was unhandled.

**Reply by pschatzmann** [reply]
The SDFAT is working perfectly: this issues are on the A2DP side...
https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/sandbox/player-audiokit_sdfat-a2dp/player-audiokit_sdfat-a2dp.ino

**Comment by TangerineDreamer** [comment]
when i incorporate audiokit.h in the player-sdfat-a2dp sketch,i have:

[W] AudioKit.h : 51 - sd_active = true
[W] AudioKit.h : 602 - Mode Button ignored because of conflict: 13 
[W] AudioA2DP.h : 302 - ==> state: Connecting
[W] AudioA2DP.h : 302 - ==> state: Connected
[W] AudioA2DP.h : 209 - is_a2dp_active -> true with 59904 bytes //-->??? what is this?

**Comment by pschatzmann** [comment]
The audio data that is currently in the buffer for a2dp when it starts the processing

**Comment by TangerineDreamer** [comment]
It's a new sanbox sketch you put in,do i upload the new reposite to make it work?

**Comment by pschatzmann** [comment]
No need to update!
It is in the sandbox because it is not working properly yet...

**Comment by TangerineDreamer** [comment]
Ok,i have enough for today,i hope you'll find what is going wrong :)

**Comment by pschatzmann** [comment]
I will fix the file issues first. This is for another day...
https://github.com/pschatzmann/arduino-audio-tools/issues/373

**Comment by TangerineDreamer** [comment]
Hi,
Just trying with the new library, have esp32 package 2.05
-audio kit sdfat player, works smoothly as usual in the speakers
-audiokit a2dp sink ,works well with smartphone,but not with audiotkit basic a2dp source (esp32 dev kit1)
-basic a2dp sink,(esp32 dev kit1) don't pair with bluetooth smartphone,don't work with any audiokit a2dp source
-basic audiokit a2dp source don't give anymore sound
-audiokit sdfat player a2dp gives this error compiling,this happened also before when i tried esp32 2.05,but not with 2.01:

Sketch uses 1347313 bytes (102%) of program storage space. Maximum is 1310720 bytes.
Global variables use 44372 bytes (13%) of dynamic memory, leaving 283308 bytes for local variables. Maximum is 327680 bytes.
Sketch too big; see https://support.arduino.cc/hc/en-us/articles/360013825179 for tips on reducing it.
text section exceeds available space in board
Compilation error: text section exceeds available space in board

Should i get the last esp32 libraries?

**Comment by pschatzmann** [comment]
Sketch uses 1347313 bytes (102%) of program storage space. Maximum is 1310720 bytes.
Just change the partition scheme to hugh app!

**Comment by TangerineDreamer** [comment]
How can i do that?

**Comment by TangerineDreamer** [comment]
ok have seen it,i'll give a try

**Comment by TangerineDreamer** [comment]
Ok,go back to esp32 package 2.02,everything now works as before but still problem with "audiokit sdfat a2dp source"

**Comment by TangerineDreamer** [comment]
Ok,problem solved,see topic #502,the working sketch is "base-player-a2dp" from the examples-basic-api.
